File: sdks/js/src/dispatch.ts
import type { Agent } from "./agent.js";
import type { BackendOptions } from "./backend.js";
import { Client } from "./client.js";
import { ConfigurationError } from "./errors.js";
import { callOf, messageOf, type InboundCall, type InboundMessage } from "./inbound.js";
import type { Session } from "./session.js";
import { Socket, number as numberOf } from "./socket.js";

/** How many calls a worker takes at once when it does not say. */
const DEFAULT_CAPACITY = 4;

/** How often the worker tells the router how it is doing. */
const DEFAULT_REPORT_EVERY_MS = 15_000;

/** How long a round trip is waited for before the last figure is left standing. */
const PING_TIMEOUT_MS = 5_000;

export interface DispatchOptions {
  /** The router to wait on. */
  client?: Client | BackendOptions;
  /**
   * How many calls to hold at once.
   *
   * The router passes over a worker that is full rather than queueing behind it, so this is
   * a promise about what this process can actually answer.
   */
  capacity?: number;
  /** How often to report load. */
  reportEveryMs?: number;
}

/** What to do with an arriving call. */
export type CallHandler = (call: InboundCall) => void | Promise<void>;

/** What to do with a message written to an agent that is not running. */
export type MessageHandler = (message: InboundMessage) => void | Promise<void>;

/**
 * Waits for inbound calls and messages, and runs a handler for each one.
 *
 * Neither arrives here first: a caller reached a Stream call over SIP, or somebody wrote in
 * a channel, and the router found out by webhook. The agent, though, runs in this process.
 * So this connects out and waits, and the router pushes work down the connection as it
 * arrives — nothing here has to be publicly reachable.
 *
 * Several workers can wait at once, in which case the work is shared between them.
 *
 * Server side only. A worker is offered other people's callers, so anything that can open
 * this socket can answer for the whole app, and the router refuses a caller that
 * authenticated as an end user's device.
 *
 * ```ts
 * const dispatch = new Dispatch({ client: { customerId: "local" } });
 *
 * dispatch.onCall(async (call) => {
 *   const agent = new Agent({ name: "John", instructions: "Be brief." });
 *   const session = await agent.answer(call);
 *   await session.wait();
 * });
 *
 * await dispatch.run();
 * ```
 */
export class Dispatch {
  readonly client: Client;
  readonly capacity: number;

  /** What the router calls this connection, for matching a log line here against one there. */
  workerId = "";

  private readonly reportEveryMs: number;
  private readonly running = new Set<Promise<void>>();
  /**
   * Which session is answering which channel.
   *
   * A channel is one conversation, so the session that answered the last message on it is
   * the one that knows what has been said and should answer the next.
   */
  private readonly answering = new Map<string, Session>();
  private onCallHandler: CallHandler | undefined;
  private onMessageHandler: MessageHandler | undefined;
  private socket: Socket | undefined;
  private pong: ((at: number) => void) | undefined;
  /** The last round trip measured, from this side, because this is the side audio crosses. */
  private latencyMs = 0;

  constructor(options: DispatchOptions = {}) {
    this.capacity = options.capacity ?? DEFAULT_CAPACITY;
    if (this.capacity < 1) {
      throw new ConfigurationError("a worker that can hold no calls cannot answer any");
    }
    this.reportEveryMs = options.reportEveryMs ?? DEFAULT_REPORT_EVERY_MS;
    this.client =
      options.client instanceof Client ? options.client : new Client(options.client ?? {});
  }

  /** How much work is being handled right now. */
  get active(): number {
    return this.running.size;
  }

  /**
   * Registers what to do with an arriving call.
   *
   * The handler runs on its own, so one long call does not stop the next from being
   * answered. What it throws is reported to the router as a call nobody took.
   */
  onCall(handler: CallHandler): this {
    this.onCallHandler = handler;
    return this;
  }

  /** Registers what to do with a message written to an agent that is not running. */
  onMessage(handler: MessageHandler): this {
    this.onMessageHandler = handler;
    return this;
  }

  /**
   * The session answering on this message's channel, started if none is.
   *
   * A channel is one conversation. The second message on it goes to the session that
   * answered the first, which is still open and knows what has been said; only a channel
   * nothing is answering calls `create`. Sessions are kept until this worker stops waiting,
   * so a conversation is not restarted between messages.
   */
  async sessionFor(
    message: InboundMessage,
    create: () => Agent | Promise<Agent>,
  ): Promise<Session> {
    const open = this.answering.get(message.channelId);
    if (open?.live) {
      return open;
    }

    const agent = await create();
    const session = await agent.reply(message);
    this.answering.set(message.channelId, session);
    return session;
  }

  /**
   * Waits for calls and messages until the router closes the connection or `signal` aborts.
   *
   * Work still being handled is waited for on the way out, because dropping a call would
   * hang up on whoever is talking.
   */
  async run(signal?: AbortSignal): Promise<void> {
    if (!this.onCallHandler && !this.onMessageHandler) {
      throw new ConfigurationError(
        "register a handler with onCall or onMessage before running",
      );
    }

    const url = await this.client.backend.socketURL("/v1/dispatch", {
      capacity: String(this.capacity),
    });
    const socket = await Socket.open(this.client.backend, url);
    this.socket = socket;

    const close = () => socket.close();
    signal?.addEventListener("abort", close, { once: true });
    // A signal that aborted while the socket was still being opened would otherwise be
    // missed, because a listener added to an aborted signal is never called.
    if (signal?.aborted) {
      close();
    }
    const reporting = setInterval(() => void this.report(), this.reportEveryMs);

    try {
      for await (const message of socket.messages()) {
        if (message instanceof Uint8Array) {
          continue;
        }

        switch (message.type) {
          case "call":
            this.answer(callOf(message));
            break;
          case "message":
            this.write(messageOf(message));
            break;
          case "ready":
            this.workerId = typeof message["worker_id"] === "string" ? message["worker_id"] : "";
            break;
          case "pong":
            this.pong?.(numberOf(message, "at"));
            break;
          default:
            break;
        }
      }
    } finally {
      clearInterval(reporting);
      signal?.removeEventListener("abort", close);
      await this.drain();
      this.answering.clear();
      socket.close();
      this.socket = undefined;
    }
  }

  /** Stops waiting. Work already being handled is still waited for by `run`. */
  stop(): void {
    this.socket?.close();
  }

  /**
   * Starts handling one call.
   *
   * Not awaited, because reading the socket is also what delivers the next call: answering
   * one caller in line would leave the next listening to a ringing phone.
   */
  private answer(call: InboundCall): void {
    const handler = this.onCallHandler;
    if (!handler) {
      return;
    }

    this.track(
      (async () => {
        try {
          await handler(call);
        } catch (cause) {
          // The router is told, so a call nobody answered shows up there rather than only
          // in this process's log.
          this.tell({
            type: "rejected",
            call_id: call.callId,
            reason: cause instanceof Error ? cause.message : String(cause),
          });
          return;
        }
        this.tell({ type: "accepted", call_id: call.callId });
      })(),
    );
  }

  /**
   * Starts handling one message, on its own for the same reason a call is.
   *
   * Nothing is reported back to the router. Accepting and rejecting are about a caller
   * waiting on a line, and there is no line here: a message nobody answered is a log line,
   * not a silence somebody is sitting in.
   */
  private write(message: InboundMessage): void {
    const handler = this.onMessageHandler;
    if (!handler) {
      return;
    }
    this.track(Promise.resolve(handler(message)).then(() => undefined));
  }

  private track(work: Promise<void>): void {
    // Settled rather than awaited, so a handler that throws does not become an unhandled
    // rejection; whether it threw is already decided by whoever created the promise.
    const held = work.catch(() => undefined);
    this.running.add(held);
    void held.finally(() => this.running.delete(held));
  }

  /**
   * Tells the router how this process is doing, on a timer.
   *
   * The router does not use any of it to choose a worker yet. It is sent so that a policy
   * which does has numbers to read, and so an operator can see which worker is under load
   * without logging into it. Only what this runtime can honestly measure is sent: host CPU
   * and memory are not portable, and an invented figure would be read as a real one.
   */
  private async report(): Promise<void> {
    await this.measure();
    this.tell({
      type: "load",
      active_agents: this.active,
      latency_ms: this.latencyMs,
    });
  }

  /** Times a round trip to the router, taken from this side because audio crosses it. */
  private measure(): Promise<void> {
    const sent = performance.now() / 1000;
    return new Promise<void>((resolve) => {
      const timer = setTimeout(() => {
        this.pong = undefined;
        resolve();
      }, PING_TIMEOUT_MS);

      this.pong = (at) => {
        clearTimeout(timer);
        this.pong = undefined;
        this.latencyMs = (performance.now() / 1000 - at) * 1000;
        resolve();
      };
      this.tell({ type: "ping", at: sent });
    });
  }

  /**
   * Sends one frame, if the socket is still there.
   *
   * A closed socket is not an error here: every one of these is something the router would
   * like to know rather than something a call depends on.
   */
  private tell(frame: Record<string, unknown>): void {
    if (this.socket?.open) {
      this.socket.send(frame);
    }
  }

  private async drain(): Promise<void> {
    await Promise.all([...this.running]);
  }
}

File: sdks/rust/src/dispatch.rs
use std::collections::{BTreeMap, HashMap};
use std::future::Future;
use std::sync::{Arc, Mutex as SyncMutex};
use std::time::Duration;

use futures_util::FutureExt;
use futures_util::future::BoxFuture;
use serde_json::{Value, json};
use tokio::sync::Mutex;
use tokio::task::JoinSet;
use tokio::time::Instant;
use tokio_util::sync::CancellationToken;

use crate::agent::Agent;
use crate::client::Client;
use crate::error::{Error, Result};
use crate::session::Session;
use crate::socket::{Frame, Incoming, SocketSender};

/// How many calls a worker takes at once when it does not say.
const DEFAULT_CAPACITY: u32 = 4;

/// How often the worker tells the router how it is doing.
const REPORT_EVERY: Duration = Duration::from_secs(15);

/// A call that arrived, as the router hands it to a worker.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct InboundCall {
    /// The Stream call the caller is already in, which is the one to join.
    pub call_id: String,
    pub call_type: String,
    /// The number they rang, which is what the agent acts from and can transfer on.
    pub called_number: String,
    /// The number they rang from, where the vendor passed it on.
    pub caller_number: String,
    /// Whatever was put on the Stream call, for a deployment that routes on its own fields.
    pub custom: BTreeMap<String, String>,
    /// When the call arrived, as the router wrote it.
    pub at: String,
}

impl InboundCall {
    fn of(frame: &Frame) -> Self {
        InboundCall {
            call_id: frame.text("call_id").into(),
            call_type: or(frame.text("call_type"), "default"),
            called_number: frame.text("called_number").into(),
            caller_number: frame.text("caller_number").into(),
            custom: frame
                .nested("custom")
                .map(|custom| {
                    custom
                        .iter()
                        .map(|(key, value)| {
                            let value = value
                                .as_str()
                                .map_or_else(|| value.to_string(), str::to_string);
                            (key.clone(), value)
                        })
                        .collect()
                })
                .unwrap_or_default(),
            at: frame.text("at").into(),
        }
    }
}

/// A message written to an agent that is not running.
///
/// One written to an agent that is already running never arrives here: the router answers
/// it from that session, because that agent knows what has been said so far.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct InboundMessage {
    /// The channel it was written in, which is the conversation to answer in.
    pub channel_id: String,
    pub channel_type: String,
    /// Who to answer as, which names the channel replies are written into.
    pub agent_id: String,
    /// The stored agent config the router matched, if it matched one.
    pub config_id: String,
    pub text: String,
    pub message_id: String,
    pub user_id: String,
    pub user_name: String,
    pub at: String,
}

impl InboundMessage {
    fn of(frame: &Frame) -> Self {
        InboundMessage {
            channel_id: frame.text("channel_id").into(),
            channel_type: or(frame.text("channel_type"), "agent"),
            agent_id: frame.text("agent_id").into(),
            config_id: frame.text("config_id").into(),
            text: frame.text("text").into(),
            message_id: frame.text("message_id").into(),
            user_id: frame.text("user_id").into(),
            user_name: frame.text("user_name").into(),
            at: frame.text("at").into(),
        }
    }
}

type CallHandler = Arc<dyn Fn(InboundCall) -> BoxFuture<'static, Result<()>> + Send + Sync>;
type MessageHandler = Arc<dyn Fn(InboundMessage) -> BoxFuture<'static, Result<()>> + Send + Sync>;

/// Waits for inbound calls and messages, and runs a handler for each one.
///
/// Neither arrives here first: a caller reached a Stream call over SIP, or somebody wrote in
/// a channel, and the router found out by webhook. So this connects out and waits, and the
/// router pushes work down the connection; nothing here has to be publicly reachable.
/// Several workers can wait at once and share the work.
///
/// Server side only: a worker is offered other people's callers.
///
/// ```no_run
/// # async fn example(client: vision_agents::Client) -> vision_agents::Result<()> {
/// use vision_agents::{Agent, Dispatch};
///
/// let dispatch = Dispatch::new(client);
/// dispatch.wait_for_call(|call| async move {
///     let session = Agent::new("support").answer(&call).await?;
///     session.wait().await;
///     Ok(())
/// });
/// dispatch.run().await
/// # }
/// ```
#[derive(Clone)]
pub struct Dispatch {
    inner: Arc<Inner>,
}

struct Inner {
    client: Client,
    capacity: u32,
    on_call: SyncMutex<Option<CallHandler>>,
    on_message: SyncMutex<Option<MessageHandler>>,
    /// Which session is answering which channel. A channel is one conversation, so the
    /// session that answered the last message on it should answer the next.
    answering: Mutex<HashMap<String, Arc<Session>>>,
    worker_id: SyncMutex<String>,
    stop: CancellationToken,
}

impl Dispatch {
    pub fn new(client: Client) -> Self {
        Dispatch::with_capacity(client, DEFAULT_CAPACITY)
    }

    /// A worker holding up to `capacity` calls at once. The router passes over a worker
    /// that is full rather than queueing behind it, so this is a promise about what this
    /// process can actually answer.
    pub fn with_capacity(client: Client, capacity: u32) -> Self {
        Dispatch {
            inner: Arc::new(Inner {
                client,
                capacity: capacity.max(1),
                on_call: SyncMutex::new(None),
                on_message: SyncMutex::new(None),
                answering: Mutex::new(HashMap::new()),
                worker_id: SyncMutex::new(String::new()),
                stop: CancellationToken::new(),
            }),
        }
    }

    /// What the router calls this connection, once it has said.
    pub fn worker_id(&self) -> String {
        self.inner.worker_id.lock().expect("worker id").clone()
    }

    /// Registers what to do with an arriving call.
    ///
    /// The handler runs as its own task, so one long call does not stop the next from being
    /// answered. An error it returns is reported to the router as a call nobody took.
    pub fn wait_for_call<F, Fut>(&self, handler: F) -> &Self
    where
        F: Fn(InboundCall) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<()>> + Send + 'static,
    {
        let handler: CallHandler = Arc::new(move |call| handler(call).boxed());
        *self.inner.on_call.lock().expect("handler") = Some(handler);
        self
    }

    /// Registers what to do with a message written to an agent that is not running. It runs
    /// as its own task, the way a call's handler does.
    pub fn wait_for_message<F, Fut>(&self, handler: F) -> &Self
    where
        F: Fn(InboundMessage) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<()>> + Send + 'static,
    {
        let handler: MessageHandler = Arc::new(move |message| handler(message).boxed());
        *self.inner.on_message.lock().expect("handler") = Some(handler);
        self
    }

    /// The session answering on this message's channel, started if none is.
    ///
    /// The second message on a channel goes to the session that answered the first, which
    /// is still open and knows what has been said; only a channel nothing is answering calls
    /// `create`. Sessions are kept until this worker stops waiting.
    pub async fn get_or_create_agent<F, Fut>(
        &self,
        message: &InboundMessage,
        create: F,
    ) -> Result<Arc<Session>>
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = Result<Agent>>,
    {
        let mut answering = self.inner.answering.lock().await;
        if let Some(open) = answering.get(&message.channel_id)
            && open.live()
        {
            return Ok(open.clone());
        }
        let session = Arc::new(create().await?.reply(message).await?);
        answering.insert(message.channel_id.clone(), session.clone());
        Ok(session)
    }

    /// Waits for calls and messages until the router closes the connection or
    /// [`Dispatch::stop`] is called.
    ///
    /// Work still being handled is waited for on the way out, because dropping a call would
    /// hang up on whoever is talking.
    pub async fn run(&self) -> Result<()> {
        let (on_call, on_message) = (
            self.inner.on_call.lock().expect("handler").clone(),
            self.inner.on_message.lock().expect("handler").clone(),
        );
        if on_call.is_none() && on_message.is_none() {
            return Err(Error::configuration(
                "register a handler with wait_for_call or wait_for_message first",
            ));
        }

        let socket = self
            .inner
            .client
            .socket(&format!("/v1/dispatch?capacity={}", self.inner.capacity))
            .await?;
        let (sender, mut receiver) = (socket.sender, socket.receiver);
        let mut running: JoinSet<()> = JoinSet::new();
        let started = Instant::now();
        let mut latency_ms = 0.0;
        let mut report = tokio::time::interval_at(Instant::now() + REPORT_EVERY, REPORT_EVERY);

        loop {
            while running.try_join_next().is_some() {}
            let incoming = tokio::select! {
                _ = self.inner.stop.cancelled() => break,
                _ = report.tick() => {
                    tell(&sender, json!({"type": "load", "active_agents": running.len(), "latency_ms": latency_ms})).await;
                    // Timed from this side, because this is the side audio crosses.
                    tell(&sender, json!({"type": "ping", "at": started.elapsed().as_secs_f64()})).await;
                    continue;
                }
                incoming = receiver.next() => incoming,
            };
            let frame = match incoming {
                Some(Ok(Incoming::Frame(frame))) => frame,
                Some(Ok(Incoming::Audio(_))) => continue,
                Some(Err(_)) | None => break,
            };

            match frame.kind() {
                "call" => {
                    let Some(handler) = on_call.clone() else {
                        continue;
                    };
                    let (call, sender) = (InboundCall::of(&frame), sender.clone());
                    running.spawn(async move {
                        let call_id = call.call_id.clone();
                        // Spawned again so a handler that panics is reported like one that failed.
                        let outcome = tokio::spawn(handler(call)).await;
                        let answer = match outcome {
                            Ok(Ok(())) => json!({"type": "accepted", "call_id": call_id}),
                            Ok(Err(error)) => json!({"type": "rejected", "call_id": call_id, "reason": error.to_string()}),
                            Err(_) => json!({"type": "rejected", "call_id": call_id, "reason": "the handler panicked"}),
                        };
                        tell(&sender, answer).await;
                    });
                }
                // Nothing is reported back for a message: there is no caller waiting on a line.
                "message" => {
                    let Some(handler) = on_message.clone() else {
                        continue;
                    };
                    let message = InboundMessage::of(&frame);
                    running.spawn(async move {
                        let _ = tokio::spawn(handler(message)).await;
                    });
                }
                "ready" => {
                    *self.inner.worker_id.lock().expect("worker id") =
                        frame.text("worker_id").into()
                }
                "pong" => {
                    latency_ms = (started.elapsed().as_secs_f64() - frame.number("at")) * 1000.0
                }
                _ => {}
            }
        }

        while running.join_next().await.is_some() {}
        let answering: Vec<Arc<Session>> = self
            .inner
            .answering
            .lock()
            .await
            .drain()
            .map(|(_, session)| session)
            .collect();
        for session in answering {
            session.close().await;
        }
        sender.close().await;
        Ok(())
    }

    /// Stops waiting. Work already being handled is still waited for by `run`.
    pub fn stop(&self) {
        self.inner.stop.cancel();
    }
}

/// Sends a frame if the socket is still there. None of these is something a call depends on.
async fn tell(sender: &SocketSender, frame: Value) {
    if sender.open() {
        let _ = sender.send(&frame).await;
    }
}

fn or(value: &str, fallback: &str) -> String {
    if value.is_empty() {
        fallback.into()
    } else {
        value.into()
    }
}
